### ISSUE TITLE
Add Keras.io Example : `Token Classification using RoBERTa`

### ISSUE THREAD

**Issue by shivance** [keras-team/keras-nlp#754]
In favor of #189

[This](https://jesusleal.io/2020/10/20/RoBERTA-Text-Classification/) huggingface tutorial can be referred !

cc : @jbischof @mattdangerw , Do you think it's required ?

**Comment by jbischof** [comment]
@shivance I think text classification we have covered in spades with https://keras.io/guides/keras_nlp/getting_started/. However, almost any other fine-tuning task beside sentence similarity would be a non-trivial extension from our guide. Any of those interesting to you?

**Comment by shivance** [comment]
@jbischof How about token classification task?

**Comment by jbischof** [comment]
That would be very interesting, thank you!

**Comment by mattdangerw** [comment]
@shivance yeah a token classification guide would be great! Could be any encoder model you could like (RoBERTa sounds good).

One option for a task would be the conll named entity recognition dataset. https://www.tensorflow.org/datasets/catalog/conll2003

**Comment by shivance** [comment]
Sure, Thanks for the input @mattdangerw .

**Comment by shivance** [comment]
An additional question, for keras-io example, the PR should be opened in [keras-io](https://github.com/keras-team/keras-io) repo right? and at `examples\nlp` location?

**Comment by mattdangerw** [comment]
Yes and yes! Feel free to ask any questions here though. We can use this for planning.

**Comment by shivance** [comment]
I was referring [Huggingface guide](https://huggingface.co/course/chapter7/2 for token classification. 
As model adds extra tokens and does subword tokenization for this task, we need to align the labels with the tokens

They used following function for that

![image](https://user-images.githubusercontent.com/51750587/219944112-95b74138-4e1b-43be-b2e3-23bad0a77bd6.png)


Huggingface Fast Tokenizer has method `word_ids()` which allows them to find the parent word of a subword token. 
![image](https://user-images.githubusercontent.com/51750587/219944199-3140bd82-2c0f-4076-a5ea-d54978763524.png)

Do we have something to tackle this problem in keras-nlp, easily and effectively. I tried to take a deeper look into implementation of `word_ids()` method, but found that it's just a python binding over rust implementation (HF tokenizers implemented in Rust). 

https://github.com/huggingface/tokenizers/blob/main/tokenizers/src/tokenizer/encoding.rs 
this is the file with class `Encoding` which is supposed to hold implementation of `word_ids()` , but I didn't find any lead.

Can you help?

**Comment by shivance** [comment]
Right now my approach is 1. trying to get parent word Id's for each token, and 2. Re use huggingface's function to align, which uses parent word ids

I am using a brute force (kindof), approach to get the parent word (later can easily be converted to IDs).
```
import re

def split(sentence):
  x = re.split(r"([^a-zA-Z0-9])", sentence)
  x = [ele for ele in x if ele != ' ' and ele !='']
  return x

def removeG(word):
  if word.startswith('Ġ'):
    return word[1:]
  return word

def find_parent_(preprocessor, sentence):

  tokens = preprocessor.tokenizer(sentence).numpy()
  sentence = split(sentence)
  #print(sentence)
  
  parent = []

  j = 0

  for i, word in enumerate(sentence):  
    tmp = ""
    tmp_par = []

    while j<len(tokens):
      #print("j = ", j)
      token = preprocessor.tokenizer.id_to_token(tokens[j])
      #print("token = ", token)
      tmp += removeG(token)
      tmp_par.append(word)

      j+=1
      #print("tmp = ", tmp, "word = ", word)
      if tmp == word:
        parent = [*parent, *tmp_par]
        #print("parent = ", parent)

        break

  return parent
```

For sentence = ['Thailand', "'", 's', 'powerful', 'military', 'thinks', 'the', 'government', 'is', 'dishonest', 'and', 'Prime', 'Minister', 'Banharn', 'Silpa', '-', 'archa', "'", 's', 'resignation']
it returns parent = ['Thailand', 'Thailand', "'", 's', 'powerful', 'military', 'thinks', 'the', 'government', 'is', 'dishonest', 'and', 'Prime', 'Minister', 'Banharn', 'Banharn', 'Banharn', 'Silpa', 'Silpa', '-', 'archa', 'archa', "'", 's', 'resignation']

Doesn't look correct. Can we do it in some better way and How can we do it more efficiently?

**Comment by shivance** [comment]
Now that I took a deeper look, Fairseq also used similar methodology to mine ( can be found in this [colab](https://colab.research.google.com/drive/1FJk35uvQM0ODs8VANmVAU2I46Shlpd-V?usp=sharing) )


https://github.com/facebookresearch/fairseq/blob/ad0e69cd99e1ff884041fbd8467d1404bd09847a/fairseq/models/roberta/alignment_utils.py#L12-L97

**Comment by TheAthleticCoder** [comment]
Hey!
I was wondering if we could use labelling spans which refer to the lengths of text in the input sequence that are assigned a label. Instead of mapping the subword tokens to the words in the sentence, we could map the subword tokens to the label spans instead.
Suppose:

Sentence: 'The Happy Cat'
Label Spans: [(0,2),(4,8),(10,12)]
Labels:[X,Y,Z] (would be the same length as the Label Spans)
Subtokens: ['GThe','GHa','Gppy','GCat']

The first loop could be through all the Subtokens and the second loop through the Label Spans. Then we could find the spans that are overlapping with the current token and assign the label to that token.

**Comment by mattdangerw** [comment]
@shivance good question! At the highest level, probably have the same feedback here as https://github.com/keras-team/keras-nlp/issues/741#issuecomment-1440813068 (squad is also a token classification problem basically).

Let's just do our preprocessing, including calculation of the labels, in python for now, in the simplest way we can. Let's focus on simple readable code, especially as this will end up as a keras.io guide. After the preprocessing is done, we can make sure training is still efficient by making sure the dataset is cached and not recomputed every epoch. Some sort of simple python routines to look for words sounds goods to me.

Some useful APIs we have today... `tokenizer.get_vocabulary()` will get you the the full list of tokens. And you can pass `dtype="string"` to a tokenizer, e.g. `tokenizer = keras_nlp.models.RobertaTokenizer.from_preset("roberta_base_en", dtype="string")` to get string subword token output.

Long term, we will definitely want some utilities for doing this, and we will want to be able to run all our preprocessing inside a TF graph. But let's start with a simple example built on top of our existing APIs, and use that to inform future improvements.

**Comment by shivance** [comment]
This work is still WIP.

**Comment by shivance** [comment]
Token classification almost done :
https://colab.research.google.com/drive/1bl2_vfp-rx67r-fMBD_HNj0yMYyWoA4p?usp=sharing

Couple of issues :
1. Facing some weird error while model fitting
2. I figured the logic to align labels with the preprocessed tokens (it works fine), but time complexity is high ( I couldn't figure out a vectorized way to do this). TC is O(128) per sample, we have 14k samples in conll2003

Can someone please help @jbischof @mattdangerw @abheesht17 ?

**Comment by shivance** [comment]
@atharvapurdue that doesn't help :(. I could read the dimension mismatch in the error message itself

Why I said it's weird? Because when i use MSE, it works fine, but not with sparse categorical cross entropy

**Comment by atharvapurdue** [comment]
@shivance , according to my understanding it works with MSE because the two losses function differently. Mean squared error is typically used for regression problems where the target variable is continuous. The output of your model for a regression problem is typically a single value or a set of continuous values. In this case, the shape of the output tensor and the shape of the target tensor are the same, which is why you did'nt encounter a shape mismatch error

While Sparse categorical Cross entropy works over probability distribution over class of the model. 
I think changing the tensor shape should have worked. Did you try it out?

**Comment by shivance** [comment]
> Let's just do our preprocessing, including calculation of the labels, in python for now, in the simplest way we can. Let's focus on simple readable code, especially as this will end up as a keras.io guide. After the preprocessing is done, we can make sure training is still efficient by making sure the dataset is cached and not recomputed every epoch. Some sort of simple python routines to look for words sounds goods to me.
> 

```
def align_labels_to_tokens(sample):
  sentence, labels = sample[2], sample[1]
  rowIndex = sample.name
  sentence_toks = sentence.split(" ")#examples[0]
  
  preprocessed_sentence = {}
  preprocessed_sentence["token_ids"] = preprocessed_mini["token_ids"][rowIndex]
  preprocessed_sentence["padding_mask"] = preprocessed_mini["padding_mask"][rowIndex]
  #preprocessor(sentence)
  tokenized_sentence = preprocessed_sentence["token_ids"]
  tokenized_sentence_to_words = []

  for tok in tokenized_sentence:
    tokenized_sentence_to_words.append(preprocessor.tokenizer.id_to_token(tok))

  parent_words = []

  """
  loop over tokenized_sentence_to_words, keep on taking words until current word 
  in sentence_toks is formed
  """

  i = 0 # pointer to sentence_toks
  j = 0 # pointer to tokenized_sentence_to_words

  current_word = ""
  buffer_cnt = 0

  while j<len(tokenized_sentence_to_words):

    if tokenized_sentence[j] <= 3:
      parent_words.append(-100)
    else:    
      current_token = tokenized_sentence_to_words[j]
      buffer_cnt += 1

      if current_token[0] == 'Ġ':
        current_token = current_token[1:]
      
      current_word += current_token

      if current_word == sentence_toks[i]:
        buffer_parent = [i] * buffer_cnt
        parent_words = [*parent_words, *buffer_parent]

        current_word = ""
        buffer_cnt = 0
        i += 1
    j+=1

  aligned_labels = []

  for i, parent_id in enumerate(parent_words):

    if parent_id>=0:
      # check if previous word had same parent
      if i > 0 and parent_id == parent_words[i-1]:
        aligned_labels.append(-100)
      else:
        aligned_labels.append(labels[parent_id])
    else:
      aligned_labels.append(-100) # special token <s>, </s>, <pad>
  
  return aligned_labels
```

**Comment by shivance** [comment]
Hey @mattdangerw were you able to find time to check my work out?

**Comment by mattdangerw** [comment]
Hi! Sorry I am always behind on my backlog of reviews.

Overall, I am hoping we can figure out something a little more concise and readable for the character offset mapping, but this works! I think the most effective thing to do here would just be to open up a PR on keras.io, and we can review with properly line by line commenting over there.

Also copying @sampathweb who might help as a reviewer here too!

**Comment by shivance** [comment]
Sure @mattdangerw , will open a PR soon.
